### ISSUE TITLE
Unable to execute pytest correctly

### ISSUE THREAD

**Issue by 2994186010** [libfuse/libfuse#441]
`sudo python3 -m pytest test/`

There is a failed test case, and the output is as following:
```
========================================================================== test session starts ==========================================================================
platform linux -- Python 3.7.3, pytest-5.0.1, py-1.8.0, pluggy-0.12.0 -- /root/miniconda3/bin/python3
cachedir: .pytest_cache
rootdir: /root/fuse1/fuse-3.6.2/build/test, inifile: pytest.ini
collected 43 items                                                                                                                                                      

test/test_ctests.py::test_write_cache[False] SKIPPED                                                                                                              [  2%]
test/test_ctests.py::test_write_cache[True] SKIPPED                                                                                                               [  4%]
test/test_ctests.py::test_notify1[True-notify_inval_inode] PASSED                                                                                                 [  6%]
test/test_ctests.py::test_notify1[True-invalidate_path] PASSED                                                                                                    [  9%]
test/test_ctests.py::test_notify1[True-notify_store_retrieve] PASSED                                                                                              [ 11%]
test/test_ctests.py::test_notify1[False-notify_inval_inode] PASSED                                                                                                [ 13%]
test/test_ctests.py::test_notify1[False-invalidate_path] PASSED                                                                                                   [ 16%]
test/test_ctests.py::test_notify1[False-notify_store_retrieve] PASSED                                                                                             [ 18%]
test/test_ctests.py::test_notify_file_size[True] PASSED                                                                                                           [ 20%]
test/test_ctests.py::test_notify_file_size[False] PASSED                                                                                                          [ 23%]
test/test_examples.py::test_hello[hello-options0-invoke_directly] PASSED                                                                                          [ 25%]
test/test_examples.py::test_hello[hello-options0-invoke_mount_fuse] PASSED                                                                                        [ 27%]
test/test_examples.py::test_hello[hello-options0-invoke_mount_fuse_drop_privileges] PASSED                                                                        [ 30%]
test/test_examples.py::test_hello[hello-options1-invoke_directly] PASSED                                                                                          [ 32%]
test/test_examples.py::test_hello[hello-options1-invoke_mount_fuse] PASSED                                                                                        [ 34%]
test/test_examples.py::test_hello[hello-options1-invoke_mount_fuse_drop_privileges] PASSED                                                                        [ 37%]
test/test_examples.py::test_hello[hello_ll-options0-invoke_directly] PASSED                                                                                       [ 39%]
test/test_examples.py::test_hello[hello_ll-options0-invoke_mount_fuse] PASSED                                                                                     [ 41%]
test/test_examples.py::test_hello[hello_ll-options0-invoke_mount_fuse_drop_privileges] PASSED                                                                     [ 44%]
test/test_examples.py::test_hello[hello_ll-options1-invoke_directly] PASSED                                                                                       [ 46%]
test/test_examples.py::test_hello[hello_ll-options1-invoke_mount_fuse] PASSED                                                                                     [ 48%]
test/test_examples.py::test_hello[hello_ll-options1-invoke_mount_fuse_drop_privileges] PASSED                                                                     [ 51%]
test/test_examples.py::test_passthrough[False-passthrough-False] FAILED                                                                                           [ 53%]

=============================================================================== FAILURES ================================================================================
_______________________________________________________________ test_passthrough[False-passthrough-False] _______________________________________________________________
Traceback (most recent call last):
  File "/root/fuse1/fuse-3.6.2/build/test/test_examples.py", line 167, in test_passthrough
    subprocess.check_call(syscall_test_cmd)
  File "/root/miniconda3/lib/python3.7/subprocess.py", line 347, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['/root/fuse1/fuse-3.6.2/build/test/../test/test_syscalls', '/tmp/tmpo0k3pysy/mnt/tmp/tmpo0k3pysy/src', ':/tmp/tmpo0k3pysy/src']' returned non-zero exit status 1.
------------------------------------------------------------------------- Captured stderr call --------------------------------------------------------------------------
  1 [create] OK
  2 [create+unlink] OK
  3 [symlink] OK
  4 [link] OK
  5 [link-unlink-link] OK
  6 [mknod] OK
  7 [mkfifo] OK
  8 [mkdir] OK
  9 [rename file] OK
 10 [rename dir] OK
 11 [rename dir loop] test_rename_dir_loop() - rename: File exists
 12 [seekdir] OK
 13 [socket] OK
 14 [utime] OK
 15 [truncate(0)] OK
 16 [truncate(13)] OK
 17 [truncate(26)] OK
 18 [truncate(126)] OK
 19 [ftruncate(0) mode: 0600] OK
 20 [ftruncate(13) mode: 0600] OK
 21 [ftruncate(26) mode: 0600] OK
 22 [ftruncate(126) mode: 0600] OK
 23 [ftruncate(0) mode: 0400] OK
 24 [ftruncate(0) mode: 0200] OK
 25 [ftruncate(0) mode: 0000] OK
 26 [open(-, O_RDONLY, 0000)] OK
 27 [open(+, O_RDONLY, 0000)] OK
 28 [open(+, O_RDWR, 0000)] OK
 29 [open(+, O_WRONLY, 0000)] OK
 30 [open(-, O_RDWR | O_CREAT, 0600)] OK
 31 [open(+, O_RDWR | O_CREAT, 0600)] OK
 32 [open(-, O_RDWR | O_CREAT | O_TRUNC, 0600)] OK
 33 [open(+, O_RDWR | O_CREAT | O_TRUNC, 0600)] OK
 34 [open(-, O_RDONLY | O_CREAT, 0600)] OK
 35 [open(-, O_RDONLY | O_CREAT, 0400)] OK
 36 [open(-, O_RDONLY | O_CREAT, 0200)] OK
 37 [open(-, O_RDONLY | O_CREAT, 0000)] OK
 38 [open(-, O_WRONLY | O_CREAT, 0600)] OK
 39 [open(-, O_WRONLY | O_CREAT, 0400)] OK
 40 [open(-, O_WRONLY | O_CREAT, 0200)] OK
 41 [open(-, O_WRONLY | O_CREAT, 0000)] OK
 42 [open(-, O_RDWR | O_CREAT, 0400)] OK
 43 [open(-, O_RDWR | O_CREAT, 0200)] OK
 44 [open(-, O_RDWR | O_CREAT, 0000)] OK
 45 [open(-, O_RDWR | O_CREAT | O_EXCL, 0600)] OK
 46 [open(+, O_RDWR | O_CREAT | O_EXCL, 0600)] OK
 47 [open(-, O_RDWR | O_CREAT | O_EXCL, 0000)] OK
 48 [open(+, O_RDWR | O_CREAT | O_EXCL, 0000)] OK
 49 [open_acc(O_RDONLY) mode: 0600 message: 'Success'] OK
 50 [open_acc(O_WRONLY) mode: 0600 message: 'Success'] OK
 51 [open_acc(O_RDWR) mode: 0600 message: 'Success'] OK
 52 [open_acc(O_RDONLY) mode: 0400 message: 'Success'] OK
 53 [open_acc(O_WRONLY) mode: 0200 message: 'Success'] OK
 54 [open(O_CREAT) in read-only directory] OK
 55 [open(O_CREAT | O_EXCL) in read-only directory] OK
 56 [open(O_CREAT | O_WRONLY) in read-only directory] OK
 57 [open(O_CREAT | O_TRUNC) in read-only directory] OK
1 tests failed
=========================================================================== warnings summary ============================================================================
/root/miniconda3/lib/python3.7/site-packages/_pytest/mark/structures.py:332
/root/miniconda3/lib/python3.7/site-packages/_pytest/mark/structures.py:332
  /root/miniconda3/lib/python3.7/site-packages/_pytest/mark/structures.py:332: PytestUnknownMarkWarning: Unknown pytest.mark.uses_fuse - is this a typo?  You can register custom marks to avoid this warning - for details, see https://docs.pytest.org/en/latest/mark.html
    PytestUnknownMarkWarning,

-- Docs: https://docs.pytest.org/en/latest/warnings.html
======================================================================== short test summary info ========================================================================
SKIPPED [2] test_ctests.py:21: not supported by running kernel
FAILED test/test_examples.py::test_passthrough[False-passthrough-False] - subprocess.CalledProcessError: Command '['/root/fuse1/fuse-3.6.2/build/test/../test/test_sys...
====================================================== 1 failed, 20 passed, 2 skipped, 2 warnings in 20.19 seconds ======================================================
```

**Comment by Nikratio** [comment]
Thanks for the report! Do you think you could work on this (or do you know someone who could)?

libfuse is a fully volunteer driven project. This means that new features and bugfixes are implemented only when someone has a personal interest in them and therefore also does the necessary
work. So if someone (that probably means you) is interested in working on this, then I'd be happy to give guidance and review patches. If not, I will close this issue in a few days (since the purpose of the issue tracker is to coordinate ongoing and planned work, rather than enumerate all the work that could possibly be done and this test does pass for many people).

**Comment by fsgeek** [comment]
I see this on my system as well and have been looking at it.  From what I can tell, it is subtle, since I don't see the failure if I run the failing test (test_syscalls) on the failing example (passthrough_ll) outside the test harness with just the debugger and test running.  I have verified that inside passthrough_ll for this failing test the size of the input file is zero, even though the test program has just written the data out to the file (test_syscalls.c in test_copy_file_range).

At this point I added an `fsync` call after the data has been written (in test_syscalls.c). Much to my surprise, this causes a _different_ test to fail because it cannot find a file:

test/test_examples.py::test_passthrough_hp[True] FAILED                                                                        [ 83%]

Traceback (most recent call last):
  File "/home/tony/projects/Araneae/src/tarantula/libfuse/build/test/test_examples.py", line 204, in test_passthrough_hp
    tst_rmdir(src_dir, mnt_dir)
  File "/home/tony/projects/Araneae/src/tarantula/libfuse/build/test/test_examples.py", line 419, in tst_rmdir
    assert name in os.listdir(mnt_dir)
AssertionError: assert 'testfile_200' in ['testfile_194', 'testfile_195', 'testfile_196', 'testfile_197', 'testfile_198', 'testfile_199']
 +  where ['testfile_194', 'testfile_195', 'testfile_196', 'testfile_197', 'testfile_198', 'testfile_199'] = <built-in function listdir>('/tmp/tmpp_v7gkrg/mnt')
 +    where <built-in function listdir> = os.listdir

I'm not sure what is causing this, but when I omit thhe passthrough_hp tests, the original failure is resolved; this fits with my theory that this may be related to the specific file system being used (I have btrfs on this test machine) and the fact that it may not implement `copy_file_range` to handle this instance (e.g., where the data is still in the cache.)

I confirmed that if I replace the `fsync` with `close/open` instead, it also resolves the issue (as expected) but the passthrough_hp test continues to fail, albeit in a _different_ spot..   The error changes slightly (the failing file name is different):

_
Traceback (most recent call last):
  File "/home/tony/projects/libfuse/build/test/test_examples.py", line 203, in test_passthrough_hp
    tst_unlink(src_dir, mnt_dir)
  File "/home/tony/projects/libfuse/build/test/test_examples.py", line 395, in tst_unlink
    assert name in os.listdir(mnt_dir)
AssertionError: assert 'testfile_201' in ['testfile_194', 'testfile_195', 'testfile_196', 'testfile_197', 'testfile_198', 'testfile_199']
 +  where ['testfile_194', 'testfile_195', 'testfile_196', 'testfile_197', 'testfile_198', 'testfile_199'] = <built-in function listdir>('/tmp/tmp3yn7j90l/mnt')
 +    where <built-in function listdir> = os.listdir

Bottom line: I don't think this is a bug in the example.  I think it either a bug in btrfs or the test program, as they clearly do not agree on the correct semantics in this case; the failing tests around the last file created in a directory not being visible immediately following suggests this may just be a btrfs behavioral issue.   I'll look at it more later, but thought I should annotate this since @Nikratio said he'd close this out if nobody was looking at it.
